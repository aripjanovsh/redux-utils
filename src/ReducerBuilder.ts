import { update } from "immupdate";
import { Reducer } from "redux";

import {
  invariant,
  isArray,
  isFunction,
  isPlainObject,
  warning,
} from "./internal/utils";
import { SubReducer } from "./types";

// eslint-disable-next-line typescript/no-type-alias,typescript/no-explicit-any
type PlainObject = { [name: string]: any };

export class ReducerBuilder<S extends PlainObject> {
  protected readonly initialState: S;

  protected readonly reducers: Array<SubReducer<S>> = [];

  public constructor(initialState: S) {
    invariant(
      isPlainObject(initialState),
      "ReducerBuilder: `initialState` expected to be a plain object.",
    );

    this.initialState = initialState;
  }

  public addSubReducer(reducer: SubReducer<S>): this {
    invariant(
      isFunction(reducer),
      "ReducerBuilder: `reducer` expected to be a function.",
    );

    this.reducers.push(reducer);

    return this;
  }

  public addActionsSubReducer(actions: string[], reducer: SubReducer<S>): this {
    invariant(
      isArray(actions),
      "ReducerBuilder: `actions` expected to be a array.",
    );

    invariant(
      isFunction(reducer),
      "ReducerBuilder: `reducer` expected to be a function.",
    );

    if (actions.length === 0) {
      warning(true, "ReducerBuilder: `actions` expected not to be empty.");
    } else {
      this.addSubReducer(
        (state, action) =>
          actions.indexOf(action.type) === -1 ? state : reducer(state, action),
      );
    }

    return this;
  }

  public addChildSubReducer<K extends keyof S>(
    childKey: K,
    reducer: SubReducer<S[K]>,
  ) {
    invariant(
      childKey != null,
      "ReducerBuilder: `childKey` expected to be defined.",
    );

    invariant(
      isFunction(reducer),
      "ReducerBuilder: `reducer` expected to be a function.",
    );

    this.addSubReducer((state, action) =>
      update(state, { [childKey]: reducer(state[childKey], action) }),
    );

    return this;
  }

  public build(): Reducer<S> {
    warning(
      this.reducers.length > 0,
      "ReducerBuilder: final reducer has no sub reducers.",
    );

    return (state = this.initialState, action) =>
      this.reducers.reduce((acc, reducer) => reducer(acc, action), state);
  }
}
