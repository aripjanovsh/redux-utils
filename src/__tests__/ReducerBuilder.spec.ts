import { ReducerBuilder } from "../ReducerBuilder";
import { mockConsole } from "./mockConsole";

interface CounterState {
  readonly counter: number;
}

function getCounterIntialState(): CounterState {
  return { counter: 0 };
}

function counterReducer(state: CounterState): CounterState {
  return { counter: state.counter + 1 };
}

describe("ReducerBuilder", () => {
  const identity = <T>(s: T): T => s;
  const consoleError = jest.fn();

  mockConsole({ error: consoleError });

  test("init validation", () => {
    expect(() => new ReducerBuilder([])).toThrowErrorMatchingSnapshot();

    expect(
      () => new ReducerBuilder(null as any),
    ).toThrowErrorMatchingSnapshot();

    expect(
      () => new ReducerBuilder(undefined as any),
    ).toThrowErrorMatchingSnapshot();

    expect(consoleError).not.toBeCalled();
  });

  test("build without sub reducers", () => {
    const builder = new ReducerBuilder({});

    builder.build();

    expect(consoleError.mock.calls).toMatchSnapshot();
  });

  test("sub reducer validation", () => {
    const builder = new ReducerBuilder({});

    expect(() =>
      builder.addSubReducer(null as any),
    ).toThrowErrorMatchingSnapshot();

    expect(() =>
      builder.addSubReducer(undefined as any),
    ).toThrowErrorMatchingSnapshot();

    expect(consoleError).not.toBeCalled();
  });

  test("sub reducer build", () => {
    const reducer = new ReducerBuilder(getCounterIntialState())
      .addSubReducer(counterReducer)
      .build();

    expect(reducer(undefined, { type: "foo" })).toEqual({ counter: 1 });
    expect(reducer({ counter: 1 }, { type: "foo" })).toEqual({ counter: 2 });

    expect(consoleError).not.toBeCalled();
  });

  test("actions reducer validation", () => {
    const builder = new ReducerBuilder({});

    expect(() =>
      builder.addActionsSubReducer(null as any, identity),
    ).toThrowErrorMatchingSnapshot();

    expect(() =>
      builder.addActionsSubReducer(undefined as any, identity),
    ).toThrowErrorMatchingSnapshot();

    expect(consoleError).not.toBeCalled();

    builder.addActionsSubReducer([], identity);

    expect(consoleError.mock.calls).toMatchSnapshot();

    expect(() =>
      builder.addActionsSubReducer(["foo"], null as any),
    ).toThrowErrorMatchingSnapshot();

    expect(() =>
      builder.addActionsSubReducer(["foo"], undefined as any),
    ).toThrowErrorMatchingSnapshot();

    expect(consoleError).not.toBeCalled();
  });

  test("actions reducer with empty actions", () => {
    const reducer = new ReducerBuilder(getCounterIntialState())
      .addActionsSubReducer([], counterReducer)
      .build();

    expect(reducer(undefined, { type: "foo" })).toEqual({ counter: 0 });
    expect(reducer(undefined, { type: "bar" })).toEqual({ counter: 0 });

    expect(reducer({ counter: 1 }, { type: "foo" })).toEqual({ counter: 1 });
    expect(reducer({ counter: 1 }, { type: "bar" })).toEqual({ counter: 1 });

    expect(consoleError.mock.calls).toMatchSnapshot();
  });

  test("actions reducer with actions", () => {
    const reducer = new ReducerBuilder(getCounterIntialState())
      .addActionsSubReducer(["foo"], counterReducer)
      .build();

    expect(reducer(undefined, { type: "foo" })).toEqual({ counter: 1 });
    expect(reducer(undefined, { type: "bar" })).toEqual({ counter: 0 });

    expect(reducer({ counter: 1 }, { type: "foo" })).toEqual({ counter: 2 });
    expect(reducer({ counter: 1 }, { type: "bar" })).toEqual({ counter: 1 });

    expect(consoleError).not.toBeCalled();
  });

  test("child reducer validation", () => {
    const builder = new ReducerBuilder(getCounterIntialState());

    expect(() =>
      builder.addChildSubReducer(null as any, identity),
    ).toThrowErrorMatchingSnapshot();

    expect(() =>
      builder.addChildSubReducer(undefined as any, identity),
    ).toThrowErrorMatchingSnapshot();

    expect(() =>
      builder.addChildSubReducer("counter", null as any),
    ).toThrowErrorMatchingSnapshot();

    expect(() =>
      builder.addChildSubReducer("counter", undefined as any),
    ).toThrowErrorMatchingSnapshot();

    expect(consoleError).not.toBeCalled();
  });

  test("child reducer", () => {
    const reducer = new ReducerBuilder(getCounterIntialState())
      .addChildSubReducer("counter", state => state + 1)
      .build();

    expect(reducer(undefined, { type: "foo" })).toEqual({ counter: 1 });
    expect(reducer(undefined, { type: "bar" })).toEqual({ counter: 1 });

    expect(reducer({ counter: 1 }, { type: "foo" })).toEqual({ counter: 2 });
    expect(reducer({ counter: 1 }, { type: "bar" })).toEqual({ counter: 2 });

    expect(consoleError).not.toBeCalled();
  });
});
