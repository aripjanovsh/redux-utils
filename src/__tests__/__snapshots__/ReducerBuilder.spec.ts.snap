// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ReducerBuilder actions reducer validation 1`] = `"ReducerBuilder: \`actions\` expected to be a array."`;

exports[`ReducerBuilder actions reducer validation 2`] = `"ReducerBuilder: \`actions\` expected to be a array."`;

exports[`ReducerBuilder actions reducer validation 3`] = `Array []`;

exports[`ReducerBuilder actions reducer validation 4`] = `"ReducerBuilder: \`reducer\` expected to be a function."`;

exports[`ReducerBuilder actions reducer validation 5`] = `"ReducerBuilder: \`reducer\` expected to be a function."`;

exports[`ReducerBuilder actions reducer with empty actions 1`] = `
Array [
  Array [
    "ReducerBuilder: final reducer has no sub reducers.",
  ],
]
`;

exports[`ReducerBuilder build without sub reducers 1`] = `
Array [
  Array [
    "ReducerBuilder: final reducer has no sub reducers.",
  ],
]
`;

exports[`ReducerBuilder child reducer validation 1`] = `"ReducerBuilder: \`childKey\` expected to be defined."`;

exports[`ReducerBuilder child reducer validation 2`] = `"ReducerBuilder: \`childKey\` expected to be defined."`;

exports[`ReducerBuilder child reducer validation 3`] = `"ReducerBuilder: \`reducer\` expected to be a function."`;

exports[`ReducerBuilder child reducer validation 4`] = `"ReducerBuilder: \`reducer\` expected to be a function."`;

exports[`ReducerBuilder init validation 1`] = `"ReducerBuilder: \`initialState\` expected to be a plain object."`;

exports[`ReducerBuilder init validation 2`] = `"ReducerBuilder: \`initialState\` expected to be a plain object."`;

exports[`ReducerBuilder init validation 3`] = `"ReducerBuilder: \`initialState\` expected to be a plain object."`;

exports[`ReducerBuilder sub reducer validation 1`] = `"ReducerBuilder: \`reducer\` expected to be a function."`;

exports[`ReducerBuilder sub reducer validation 2`] = `"ReducerBuilder: \`reducer\` expected to be a function."`;
